File: frontend/src/types/payments.ts
export type PaymentType = 'normal' | 'reversal';

export interface PaymentRead {
  amount: string;
  created_at: string;
  id: string;
  operator_id: string | null;
  payment_date: string;
  payment_no: string;
  payment_type: PaymentType;
  remark: string | null;
  reverses_payment_id: string | null;
  sub_project_id: string;
  updated_at: string;
}

export interface PaymentListRead {
  items: PaymentRead[];
  page: number;
  page_size: number;
  total: number;
}

export interface PaymentListQuery {
  page?: number;
  pageSize?: number;
  paymentType?: PaymentType;
}

export interface PaymentCreatePayload {
  amount: string;
  confirmOverBudget?: boolean;
  files: File[];
  overBudgetReason?: string | null;
  paymentDate: string;
  remark?: string | null;
  subProjectId: string;
}

export interface PaymentReversePayload {
  remark: string;
  reversesPaymentId: string;
  subProjectId: string;
}

export const PAYMENT_TYPE_OPTIONS = [
  { label: '全部类型', value: '' },
  { label: '正常 normal', value: 'normal' },
  { label: '红冲 reversal', value: 'reversal' },
] as const;

export const PAYMENT_TYPE_LABELS: Record<PaymentType, string> = {
  normal: '正常 normal',
  reversal: '红冲 reversal',
};

File: backend/app/schemas/payments.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.payments import PaymentType


class PaymentVoucherRead(BaseModel):
    id: UUID
    payment_id: UUID
    document_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaymentVoucherDocumentRead(BaseModel):
    id: UUID
    doc_type: str
    file_name: str
    display_name: str
    file_size: int
    version: int
    is_deleted: bool
    uploader_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaymentVoucherWithDocumentRead(PaymentVoucherRead):
    document: PaymentVoucherDocumentRead


class PaymentRead(BaseModel):
    id: UUID
    payment_no: str
    sub_project_id: UUID
    amount: Decimal
    payment_date: date
    remark: str | None
    payment_type: PaymentType
    reverses_payment_id: UUID | None
    operator_id: UUID | None
    vouchers: list[PaymentVoucherWithDocumentRead] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaymentListRead(BaseModel):
    items: list[PaymentRead]
    page: int
    page_size: int
    total: int
